Implement edit & upgrade subworkflow buttons in workflow editor.

jmchilton opened the pull request:
... building on the enhancements to new workflow refactoring API. This PR is really about the refactoring API and improving its structure but those two buttons are big usability wins for people using subworkflows so I changed the title of this PR. 

<img width="306" alt="Screen Shot 2021-01-04 at 12 35 13 AM" src="https://user-images.githubusercontent.com/216771/103504282-f8822500-4e24-11eb-96dc-4f80923906a9.png">

<img width="329" alt="Screen Shot 2021-01-04 at 12 35 19 AM" src="https://user-images.githubusercontent.com/216771/103504292-fc15ac00-4e24-11eb-9053-4aeba6712c50.png">

<img width="823" alt="Screen Shot 2021-01-04 at 12 36 41 AM" src="https://user-images.githubusercontent.com/216771/103504294-fddf6f80-4e24-11eb-846d-3c00a136fd60.png">


Workflow Refactoring API improvements:
- ~Builds on  #11016~ (merged)
- Includes https://github.com/galaxyproject/galaxy/pull/11041
- Fixes bug where the refactor API would create a new subworkflow for each update to an existing workflow.
- Fixes older bug from start of this release cycle where license wouldn't appear in export workflows.
- Implement five new refactoring actions ``fill_defaults``, ``fill_step_defaults``, ``upgrade_subworkflow``, ``upgrade_tool`` (this one is a bit beta but it isn't used in the GUI), ``update_output_label``. 
- All sorts of new tests to clarify test new actions and clarify how the API deals with missing steps, incomplete step state, etc...
- All sorts of new tests to clarify catching and reporting potential issues during subworkflow upgrades.
- Rework the workflow API response to include typed, structured messages in addition to a workflow. This structure is all important to record and store in the database with each version I think (this is outlined in #11005 and important for #9166. Four potential error types are captured:
   - ``tool_version_change``
   - ``tool_state_adjustment``
   - ``connection_drop_forced``
   - ``workflow_output_drop_forced``
- Comprehensive reworking of the ``dry_run`` option to the API. It now works and returns the workflow as it would be if it were modified according to the request. This includes a bunch of new tests that ensure the database isn't changing in response to ``dry_run`` requests.
- Use the new and improved ``dry_run`` option and the typed messages to report potential problems with refactorings in the UI and have the user confirm before up upgrading workflows.

jmchilton committed:
integration test for tool state handling around fill_defaults option

jmchilton committed:
Ensure the workflow refactor API works around incomplete tool state information.

Test cases and one small fix to ensure this.

This should allow the API (or at least many of the actions) to work without knowledge of a toolbox potentially and I think it will allow use to track the tool state upgrades the workflow editor does automatically more formally.

jmchilton committed:
wf refactor API - add fill_step_defaults action

jmchilton committed:
Rework refactoring API for better separation of concerns.

Thinner controller, pydantic model to separate options available for manager (which has to render a response) from those of the refactoring executor.

jmchilton committed:
Rework refactor responses for structured messages...

jmchilton committed:
wf refactor API - add fill_defaults action

Fills step defaults on all tool steps.

jmchilton committed:
Fix bug where subworkflows would be duplicated by refactor API.

jmchilton committed:
wf refactor API - add upgrade_subworkflow action

jmchilton committed:
wf refactor API - add upgrade_tool action

jmchilton committed:
Edit and upgrade subworkflow buttons in workflow editor.

jmchilton committed:
refactor action - upgrade_subworkflow - report dropped inputs

jmchilton committed:
Use existing workflow_manager to load subworkflows in wf modules.

jmchilton committed:
Implement much more complete dry_run workflow refactor API.

jmchilton committed:
bugfix - license not in exported workflow

jmchilton committed:
Drop unreachable showWarnings method in workflow editor.

It was added in https://github.com/galaxyproject/galaxy/commit/c922711612f217d0d780581b3ecaf5cfa3397f77 which didn't have the exception handler. showWarnings only shows a modal if the response is JSON with 'errors' in it, I've searched the model/__init__.py, managers/workflows.py, api/workflows.py, and controller/workflow.py and the only response that would include that is in the save_workflow_as. the saveAs function though does not use this function.

jmchilton committed:
wf refactor API - implement update_output_label action

jmchilton committed:
wf refactor API - subworkflow upgrade - dropping connections

jmchilton committed:
subworkflow upgrade - catch and record dropped workflow output labels

jmchilton committed:
Better separation of concerns around client interactions with refactoring API.

jmchilton committed:
wf refactor API - fix dry_run=True subworkflow upgrades

objects were slipping into sa_session in very subtle ways

jmchilton committed:
Display potential issues with refactorings for confirmation.

jmchilton committed:
Rename response key actions_executed -> action_executions

When dry_run=True actions_executed feels mislabeld.

jmchilton committed:
Better wording on upgrade subworkflow tooltip.

mvdbeek commented on lib/galaxy/managers/workflows.py:
```diff
@@ -403,23 +408,29 @@ def update_workflow_from_raw_description(self, trans, stored_workflow, raw_workf
+            workflow.stored_workflow = stored_workflow
+            stored_workflow.latest_workflow = workflow
+        else:
+            stored_workflow = model.StoredWorkflow()  # detached
```
I don't think this is a good way to build detached objects (we are using the same trick for transient collections in the CWL branch, and they did actually get persisted) ... they will be attached to a session as soon as you add something coming from a session with a bidirectional relationship. I am also not confident that our session handling prevents a commit occurring elsewhere to affect this. I don't think it's a blocker since I didn't see this in my limited testing, but maybe a better strategy would be using a pydantic model here ?

jmchilton replied:
I did extensive testing - I found all the places for these actions at least and there were some tricky ones. But in the dry run tests, I flushed the session, performed the actions, flush, and then made sure none of the workflow objects in the database were touched. If we do dry run tests for each new refactoring I think this is fine.

I really think the biggest problem is just the lib/galaxy/workflow/modules.py must injects a bunch of random state onto workflow steps that isn't defined at the model layer. If we had a clean module layer and stateless modules I don't think this would be so bad. Pydanic models that mirror the sqlite ones would be fine also - we probably should have a best practice for doing that.

jmchilton replied:
xref https://github.com/galaxyproject/galaxy/pull/11032/files#diff-346d4ccd160634059af0bbdd140037c890f4974b57cf10ee1f88eff0d1576f2fR727

mvdbeek commented on lib/galaxy/webapps/galaxy/controllers/workflow.py:
```diff
@@ -646,13 +646,20 @@ def delete(self, trans, id=None):
+                stored_workflow = workflow.stored_workflow
+                self.security_check(trans, stored_workflow, True, False)
+                stored_workflow_id = trans.security.encode_id(stored_workflow.id)
+                return trans.response.send_redirect(f'{url_for("/")}workflow/editor?id={stored_workflow_id}')
```
This will always load up the latest version of the stored workflow ... is that what you intended ? I tried to fix something similar here: https://github.com/galaxyproject/galaxy/pull/10976/commits/481aecebd4ca7abac1271a848c8e0777e6c64482

jmchilton replied:
Yeah - the GUI says "load the subworkflow at the latest version" so that is what is happening here.

Loading an older version in the editor would be bad because you might save over more recent changes. I think having a read-only view of the workflow from the editor though would be awesome - so users can see what is happening in the subworkflow but that wasn't the intention here.

mvdbeek commented on lib/galaxy/workflow/refactor/schema.py:
```diff
@@ -166,20 +167,53 @@ class UpdateReportAction(BaseAction):
+class UpgradeSubworkflowAction(BaseAction):
+    action_type: Literal['upgrade_subworkflow']
+    step: step_reference_union
+    # should be decoded before stuffing it into the database...
```
Is that a todo ?

jmchilton replied:
It is a TODO when we start storing these actions in the database along with the new Workflow object they produce - but this PR doesn't dump this stuff into the database so it isn't a TODO we can act on right now. That ongoing goal in tracked in #11005.

jmchilton replied:
I've revised the comment a bit to clarify this.

mvdbeek commented on lib/galaxy/workflow/refactor/schema.py:
```diff
@@ -190,6 +224,47 @@ class UpdateReportAction(BaseAction):
+class RefactorActionExecutionMessage(BaseModel):
+    message: str
+    message_type: RefactorActionExecutionMessageTypeEnum
+    # messages don't have to be bound to a step, but if they are should
```
```suggestion
    # messages don't have to be bound to a step, but if they are they should
```

mvdbeek commented on lib/galaxy/workflow/refactor/schema.py:
```diff
@@ -190,6 +224,47 @@ class UpdateReportAction(BaseAction):
+    # in the case of tool inputs (e.g. 'cond|repeat_0|input')
+    input_name: Optional[str]
+
+    # messages don't have to be bound to a step with inputs, but if they are
```
I really like those comments. Can those become descriptions ?

jmchilton replied:
Done? It is much harder to read inline now but they will be pretty in an OpenAPI schema someday.

mvdbeek replied:
Thanks!

mvdbeek reviewed (commented):
Very exciting, this all looks very solid. Unfortunately it failed upgrading my first subworkflow with:
<img width="767" alt="Screenshot 2021-01-06 at 11 13 51" src="https://user-images.githubusercontent.com/6804901/103756999-6cccec00-5010-11eb-8802-6785be914a92.png">

Digging through what might be going on now ...

mvdbeek committed:
Serialize RefactorResponse using safe_dumps

Workflows have dictionaries with integer keys, which pydantic doesn't coerce to strings.
Integer object keys aren't valid JSON, so the client fails.

jmchilton committed:
Convert refactor message comments to field descriptions.
